src/main/java/org/jitsi/videobridge/cc/vp8/VP8FrameMap.java

```diff
@@ -271,4 +172,127 @@ public boolean isNewFrame()
+            int index = indexTracker.interpret(pictureId);
+            ArrayTracker<VP8Frame>.Container c = super.getContainer(index);
+            if (c == null)
+            {
+                return null;
+            }
+            return c.getItem();
+        }
+
+        public boolean insert(int pictureId, VP8Frame frame)
+        {
+            int index = indexTracker.update(pictureId);
+            return super.insertItem(frame, index);
+        }
+
+
+        public VP8Frame findBefore(VP8Frame frame, Predicate<VP8Frame> pred)
+        {
+            int index = indexTracker.interpret(frame.getPictureId());
+            ArrayTracker<VP8Frame>.Container c =
+                super.findContainerBefore(index, pred::test);
+            if (c == null)
+            {
+                return null;
+            }
+           return c.getItem();
+        }
+
+        public VP8Frame findAfter(VP8Frame frame, Predicate<VP8Frame> pred)
+        {
+            int index = indexTracker.interpret(frame.getPictureId());
+            ArrayTracker<VP8Frame>.Container c =
+                super.findContainerAfter(index, pred::test);
+            if (c == null)
+            {
+                return null;
+            }
+            return c.getItem();
+        }
+
+        /** Like Rfc3711IndexTracker, but for picture IDs (so with a rollover
+         * of 0x8000).
+         */
+        private static class PictureIdIndexTracker
+        {
+            private static boolean isOlderThan(int a, int b)
+            {
+                return Vp8Utils.getExtendedPictureIdDelta(a, b) < 0;
+            }
+
+            private static boolean isNewerThan(int a, int b)
+            {
+                return Vp8Utils.getExtendedPictureIdDelta(a, b) > 0;
+            }
+
+            private static boolean rolledOverTo(int a, int b)
+            {
+                return isOlderThan(a, b) && b < a;
+            }
+
+            private int roc = 0;
+            private int highestSeqNumReceived = -1;
+
+            private int getIndex(int seqNum, boolean updateRoc)
+            {
+                if (highestSeqNumReceived == -1) {
+                    if (updateRoc) {
+                        highestSeqNumReceived = seqNum;
+                    }
+                    return seqNum;
+                }
+
+                int v;
+
+                if (rolledOverTo(seqNum, highestSeqNumReceived))
+                {
+                    v = roc - 1;
+                }
+                else if (rolledOverTo(highestSeqNumReceived, seqNum))
+                {
+                    v = roc + 1;
+                    if (updateRoc)
+                        roc = v;
+                }
+                else {
+                    v = roc;
+                }
+
+                if (updateRoc && isNewerThan(seqNum, highestSeqNumReceived))
+                {
+                    highestSeqNumReceived = seqNum;
+                }
+
+                return 0x8000 * v + seqNum;
```

Comment:
This will not work as expected if the sender uses only 7-bit PictureIDs. I don't think any implementations actually do this, but it is explicitly allowed in RFC7741: `Having sent a 7-bit PictureID with all bits set to 1, the sender may either wrap the PictureID to 0 or extend to 15 bits and continue incrementing.`

Can you think of a way to detect it if it happens, so we can log?

Reply:
I have some vague recollection that Chrome actually insists on 15-bit PictureIDs (i.e. even if one packet uses 7-bit, it expects the packet after 0x7f to be a 15-bit 0x80).  But now I can't find where this is documented.

Probably we can just treat 7-bit packets as though they didn't include a picture ID at all, since they're optional.  (Which leads to the point that we don't handle that case, currently.)

Reply:
Possibly streams without picture IDs should just be handled by the generic tracker?  I'm not sure how much we care about this case, because all the browsers we support will generate them.

Reply:
Okay, if wrapping at 0x7f breaks Chrome it won't be that hard to discover the probem.

> Probably we can just treat 7-bit packets as though they didn't include a picture ID at all

This seems more likely to cause problems in practice (when senders wrap at 0x7fff but encode small values as 7-bit)